feat: add filter types for multiselect field

southball opened the pull request:
## Change Summary

Add filters "contains all of", "contains any of", "does not contain all of" and "does not contain any of" for multiselect field (although they are enabled for all field types).
Resolves #534.

## Change type

- [X] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Tested locally with multiselect field.

## Additional information / screenshots (optional)

- I did a quick search for `'eq'` in the entire codebase and added the new operators to where it seems is needed. If there seems to be missing implementations for the new operators, please kindly let me know.

wingkwong reviewed (changes requested):
Trying the examples in the issue post. The following cases are failing.

Test Case 2:

where `tags` `contains any of` `Foo, Bar`
Expected Result: rows ❷, ❸, ❺ and ❻
Your Result: rows ❷, ❺ and ❻

Test Case 3:

where `tags` `contains all of` `Foo, Bar`
Expected Result: rows ❺ and ❻
Your Result: empty

Test Case 4:

where `tags` `is equal` `Foo, Bar`
Expected Result: rows ❺
Your Result: empty

Test Case 8:

where `tags` `does not contain any of` `Quux`
Expected Result: rows ❶, ❷, ❸ and ❺
Your Result: rows ❷, ❸ and ❺

wingkwong reviewed (commented):
besides, these new operators should be only available when filtering MultiSelect.

southball commented:
Test case 8 is a problem. I will take a look at it later.

Regarding test case 2, 3, and 4, there is a problem that there is a space after the `,`. A sensible solution would be to trim each part of the query, however it is possible that one of the value of multiselect can contain leading and/or trailing whitespaces. Is it alright to ignore such cases and pick the solution of trimming the split query?

wingkwong commented:
We may also think about when a tag contains `,`, e.g. `Foo, Bar` and the query is like `Foo, Bar`.

pranavxc commented on packages/nocodb/src/lib/meta/helpers/webhookHelpers.ts:
```diff
@@ -72,6 +72,26 @@ export async function validateCondition(filters: Filter[], data: any) {
+          (data[field]?.split(',') ?? []).includes(item)
+        );
+        break;
+      case 'nallof':
+        res = !(filter.value?.split(',') ?? []).every((item) =>
+          (data[field]?.split(',') ?? []).includes(item)
+        );
```
`notallof` - I guess it should check all of them are not matching, right now it will be true even if one is not matching.

```
  res = (filter.value?.split(',') ?? []).every((item) =>
          !(data[field]?.split(',') ?? []).includes(item)
        );

```

The same applies for `nanyof`

```
res = (filter.value?.split(',') ?? []).some((item) =>
          !(data[field]?.split(',') ?? []).includes(item)
        );
```

Correct me if I'm wrong

southball replied:
Let us take a look at "does not contain any of" (`nanyof`) first. Consider the filter `nanyof Foo,Bar` and two tag values:

1. Foo,Bar,Quux
2. Foo,Quux

Both 1, 2 should *not* be shown, because they contain some of the items in the filter.

Let us then take a look at the implementation.

```ts
      case 'nanyof':
        res = !(filter.value?.split(',').map((item) => item.trim()) ?? []).some((item) =>
          (data[field]?.split(',') ?? []).includes(item)
        );
        break;
```

In this case, `(filter.value?.split(',').map((item) => item.trim()) ?? [])` evaluates to `["Foo", "Bar"]` and so the `some` evaluates to `true` in both case. Which means that `res` is false, and the rows will not be shown.

So I *think* the implementation of `nanyof` is correct, but I am not 100% sure. What do you think?

pranavxc replied:
Now it makes sense :+1: , you can update the docs part as well to make it clear - https://github.com/nocodb/nocodb/blob/develop/packages/noco-docs/content/en/developer-resources/rest-apis.md#comparison-operators

southball commented:
It has taken me some time to get to this PR, but I have rebased the PR and added some fixes.

- Between the time that this PR is made and the rebase, it became impossible to type in spaces in the filter. However, trimming of string is still added. It should pass all test cases now.
- The filters are only now shown if the field `uidt` is `MultiSelect`.

southball commented on packages/nc-gui/components.d.ts:
```diff
@@ -90,6 +90,7 @@ declare module '@vue/runtime-core' {
     LogosMysqlIcon: typeof import('~icons/logos/mysql-icon')['default']
     LogosPostgresql: typeof import('~icons/logos/postgresql')['default']
     LogosRedditIcon: typeof import('~icons/logos/reddit-icon')['default']
+    LogosSnowflakeIcon: typeof import('~icons/logos/snowflake-icon')['default']
```
I think ESLint added this line, which should be alright?

wingkwong replied:
Yes. This was introduced from another PR. Maybe the author didn't commit this change.

southball commented on packages/nc-gui/utils/filterUtils.ts:
```diff
@@ -35,6 +35,26 @@ export const comparisonOpList = [
+  {
+    text: 'contains all of',
+    value: 'allof',
+    types: ['MultiSelect'],
```
Added a field here, so we can have more uidt-specific filters in the future.

wingkwong replied:
may use `UITypes.MultiSelect`

southball commented on packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/conditionV2.ts:
```diff
@@ -306,6 +306,46 @@ const parseConditionV2 = async (
+            ) {
+              qb = qb.where(condition);
+            } else {
+              qb = qb.whereNot(condition).orWhereNull(field);
```
Added `orWhereNull(field)` for the last test. Things behave unexpectedly for our usecase when `null` is involved.

southball commented:
Seems that CI is broken in this PR. Is rebasing this commit enough to fix the problem?

wingkwong commented:
@southball Please rebase. Seems the tests failed not because of your changes.

southball commented:
Took a bit of time, but rebased. Before rebase the develop branch seems to have passing tests, so if there are no problems this should pass I guess.

wingkwong commented on packages/nocodb/src/lib/meta/helpers/webhookHelpers.ts:
```diff
@@ -72,6 +72,26 @@ export async function validateCondition(filters: Filter[], data: any) {
       case 'notnull':
         res = data[field] !== null;
         break;
+      case 'allof':
+        res = (filter.value?.split(',').map((item) => item.trim()) ?? []).every((item) =>
+          (data[field]?.split(',') ?? []).includes(item)
+        );
+        break;
+      case 'anyof':
+        res = (filter.value?.split(',').map((item) => item.trim()) ?? []).some((item) =>
+          (data[field]?.split(',') ?? []).includes(item)
+        );
+        break;
+      case 'nallof':
+        res = !(filter.value?.split(',').map((item) => item.trim()) ?? []).every((item) =>
+          (data[field]?.split(',') ?? []).includes(item)
+        );
+        break;
+      case 'nanyof':
+        res = !(filter.value?.split(',').map((item) => item.trim()) ?? []).some((item) =>
+          (data[field]?.split(',') ?? []).includes(item)
+        );
+        break;
```
may move (data[field]?.split(',') ?? []) out of .map since the value won't be changed in each iteration.

wingkwong commented on packages/nc-gui/components/smartsheet/toolbar/ColumnFilter.vue:
```diff
@@ -213,7 +218,12 @@ defineExpose({
+                v-for="compOp in getApplicableFilters(filter.fk_column_id)"
+                :key="compOp.value"
+                :value="compOp.value"
+                class=""
```
may remove it if there's no class name

wingkwong commented on packages/nc-gui/components/smartsheet/toolbar/ColumnFilter.vue:
```diff
@@ -100,6 +100,11 @@ watch(
 )
 
+const getApplicableFilters = (id?: string) => {
+  const colType = (meta.value?.columnsById as Record<string, ColumnType>)?.[id ?? '']?.uidt
```
in which case `id` would be not required?

pranavxc replied:
if user is not selected a column id will be undefined/null
